File: lc_soln/0020_valid_parentheses/solution.py
class Solution:
    def isValid(self, s: str) -> bool:
        matching = {")": "(", "]": "[", "}": "{"}
        left = {"(", "[", "{"}
        stack = []
        for c in s:
            if c in left:
                stack.append(c)
                continue
            if not stack or stack.pop() != matching[c]:
                return False
        return not stack

File: lc_soln/0020_valid_parentheses/solution.cs
public class Solution
{
    public bool IsValid(string s)
    {
        var stack = new Stack<char>();
        foreach (char c in s)
        {
            if (c == '(' || c == '[' || c == '{')
            {
                stack.Push(c);
                continue;
            }
            if (stack.Count == 0)
            {
                return false;
            }
            char o = stack.Pop();
            if (!((o == '(' && c == ')') || (o == '[' && c == ']') || (o == '{' && c == '}')))
            {
                return false;
            }
        }
        return stack.Count == 0;
    }
}
